clears authentications when logout is clicked

leggsimon opened the pull request:
### Description
This PR allows a user to “log out” and the app forgets the state of their previously used api keys.

### Motivation and Context

Fixes #4641. Fixes #4904. 

When the logout button is clicked in the authorise popup the request was being made to the redux store to clear the key. However if the user clicked “Authorise” it sends what ever is in the component’s state and populates the redux store with that. Since, on logout, the state in the component wasn’t being cleared if you just clicked “Authorise” it would resend the same value it had before logging out.

### How Has This Been Tested?
Manually as well as adding 2 cypress tests to cover the “normal” user journey and the new bug. I ensured that the tests failed before putting in my fix.

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

### My PR contains... 
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [x] If yes to above: I have taken care to cover edge cases in my tests. **To my knowledge I can’t think of any but am happy to add more**
- [ ] All new and existing tests passed.

shockey commented on test/e2e-cypress/tests/bugs/4641.js:
```diff
@@ -0,0 +1,49 @@
+describe("#4641: The Logout button in Authorize popup not clearing API Key", () => {
```
I'd like to see a third test case that covers the following scenario.

Given this definition:

```yaml
openapi: 3.0.0
paths:
  /:
    get:
      description: ok
      security:
      - securityA: []
      - securityB: []
      responses: 
        200: 
          description: ok
components:
  securitySchemes:
    securityA:
      type: apiKey
      name: SecurityA
      in: header
    securityB:
      type: apiKey
      name: SecurityB
      in: header
```

1. I authorize securityA and securityB.
2. I logout and then attempt to reauthorize securityB (idea being: the data will be blank, so reauthorization won't happen).
3. I execute my operation.
4. **Assert that securityA was attached to the request, and securityB was not.**

My rationale here is that the data-clearing code is in the Auths component, which has the capability to wipe out more than one authorization's data, so we should have a defensive test against that happening.

leggsimon replied:
done in a0b8f4f

shockey commented on test/e2e-cypress/static/documents/bugs/4641.yaml:
```diff
@@ -0,0 +1,34 @@
+    email: support@demo.io
+  version: 1.0.0
+servers:
+  - url: 'https://httpstat.us/'
```
Cypress tests need to avoid relying on external services - if `httpstat.us` is ever offline, this test will fail because Swagger UI will show `TypeError: failed to fetch` instead of the response content.

If you just drop the `servers` block, the request will just hit the test server and you'll get back a 404, which should work for your purposes since you only care about what's in the request 😄

leggsimon replied:
done in 3ef208e

leggsimon commented:
Cool, that sounds totally reasonable. Didn’t know you could do that, I’ll update them

leggsimon commented:
@shockey I’ve added a new test based on your suggestion and also changed the tests so that it’s checking the request headers being sent (this is also way faster).

I’ve also refactored these tests slightly to extract some of the common things like filling out the auth field and clicking authenticate. Personally I think it’s slightly more readable but am happy to change it back. I know it’s not everyones cup of tea.

leggsimon commented on test/e2e-cypress/tests/bugs/4641.js:
```diff
@@ -0,0 +1,49 @@
+      .click()
+      .wait(2000) // wait for response
+      .get(".curl")
+      .should("contain", "api_key: my_api_key")
```
done in 3ef208e

leggsimon commented:
@shockey any possibility of a re-review?

shockey commented on test/e2e-cypress/tests/bugs/4641.js:
```diff
@@ -0,0 +1,98 @@
+      })
+  })
+
+  it("should not only forget the value of the auth the user logged out from", () => {
```
```suggestion
  it("should only forget the value of the auth the user logged out from", () => {
```

leggsimon commented:
Appreciate it

shockey commented:
thanks for the contribution, @leggsimon!
